### ISSUE TITLE
Conditional field serialization - ISO8583

### ISSUE THREAD

**Issue by nullpainter** [jefffhaynes/BinarySerializer#187]
I'm wishing to use BinarySerializer to both serialize and deserialize [ISO8583](https://en.wikipedia.org/wiki/ISO_8583) messages. These messages contain bitmaps to indicate whether specific fields are present or not. 

I have created custom `IBinarySerializable` bitmap properties, which work well for serialization of the bitmaps themselves. 

The difficulty I'm facing is that ISO8583 messages inherently have a dynamic length. If a property isn't present, it shouldn't consume bytes in the message.

`SerializeWhenAttribute` won't work for me, as this refers to the state of other serialized properties. 

Setting a property to null just results in `n` `0` bytes written, where `n` is the value defined in `FieldLength`.

All of my fields are currently custom. I have tried leaving `FieldLength` off  entirely and conditionally serializing based on another property, but then buffers aren't correctly sized inside `IBinarySerializable`.

I also tried setting a `ConverterType` for `FieldLength`, however the associated `IValueConverter` is only called on deserialization, not serialization.

In summary:

1. Is there a way for to not serialize a field but to also not serialize `FieldLength` bytes?
2. If not, is there a way to programmatically set `FieldLength` inside `IBinarySerializable`?

**Comment by nullpainter** [comment]
I anticipate a similar issue with `LLVAR` and `LLLVAR` data types, where the first 2-3 bytes contain the field length. Is there a reason why `ConverterType` on `FieldLength` isn't being called on serialization? I _think_ this may solve all of my problems:

1. For optional fields, I can set the field length to zero if the field is null
2. For `LLVAR` and `LLLVAR`, I can set the field length based on the property type and initial bytes

**Comment by nullpainter** [comment]
After all of that, I have created my own serializer for this. 

Whereas BinarySerializer could work, ISO8859 uses a sufficiently different approach that makes it less error-prone to do the serialization myself. For example, all integers and longs are BCD, all fields of a particular ISO data type will have the same length, all LLVAR fields will have the same length calculation etc.

So, feel free to address my questions if you wish - but I will probably not use BinarySerializer for this particular project. Until next time!

**Comment by jefffhaynes** [comment]
Sorry I'm just getting to this.  Maybe I'm missing something, but can you simply wrap the length and value in a field that is conditionally serialized?  That format looks fiendishly complicated so it may be that there is something breaking there!

**Comment by nullpainter** [comment]
It's been a while so I can't remember the exact issues I was facing, although it may be related to the length being dynamic. My custom serializer works a charm and has been a blessing in disguise, as it's allowed me to hook in custom validations, pretty printing etc. 

It's a pity I can't release it as it's commercial. The format is indeed fun, but straightforward once you get your head around it!

**Comment by jefffhaynes** [comment]
Glad to hear it!
